### PR TITLE
Remove defined marker

### LINKED ISSUE #2351: Remove Defined marker
After #2249, we can modify the `SideCondition` to keep a set of all defined terms. Instead of looking for the `Defined` marker, we can look in the `SideCondition`.

---

- [x] Add a `HashSet` of defined `TermLike`s to `SideCondition`. (appetite: S)
- [x] Implement a function `assumeDefined :: TermLike variable -> SideCondition variable` which assumes the definedness of all sub-terms that are implied by the original term. In other words, if `t` is the original term, we assume a sub-term `s` is defined if `\ceil(t) \implies \ceil(s)`. Take care with Map and Set to minimize storage! If the size of a Map or set is `n`, we only want `n^2` entries and not `n!`. (appetite: M)
- [x] Implement a function `isDefined :: SideCondition variable -> TermLike variable -> Bool`. Again, take care with Map and Set. (appetite: M)
- [ ] Remove `Defined`. Replace `makeDefined` with `assumeDefined`. Use `isDefined` instead of checking pattern attributes. (appetite: M)

### PR DESCRIPTION
**Pull request by ana-pantilie**
Fixes #2351

- removes `TermLike.Defined` entirely, definedness assumptions and checks are replaced by `assumeDefined` and `isDefined`
- makes the `SideCondition` available:
    - to the builtin evaluators
    - when matching
    - in the SMT translator/evaluator
- fixes both `assumeDefined` and `isDefined` to not ignore the values of maps

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

**Comment by ana-pantilie** [comment]
@andreiburdusa @ttuegel Sorry, I realized I forgot to add a description to the PR. This made me notice that a few changes could be cleaned up a little and that I was missing some tests. I also fixed the conflicts with master.

File: kore/src/Kore/Step/Simplification/AndTerms.hs [hunk 1]
```diff
@@ -310,7 +296,7 @@ equalAndEquals
     -> TermLike RewritingVariableName
     -> MaybeT unifier (Pattern RewritingVariableName)
 equalAndEquals first second =
-    if unDefined first == unDefined second then
+    if first == second then
         -- TODO (thomas.tuegel): Preserve defined and simplified flags.
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/AndTerms.hs:300]:
```suggestion
        -- TODO (thomas.tuegel): Preserve simplified flags.
```

File: kore/src/Kore/Step/Simplification/Exists.hs [hunk 2]
```diff
@@ -195,9 +195,14 @@ makeEvaluate sideCondition variables original = do
...
-                    variable
-                    normalized { Conditional.substitution = boundSubstitution }
+                matched <-
+                    lift
+                    $ matchesToVariableSubstitution
+                        sideCondition
+                        variable
+                        normalized
+                            { Conditional.substitution = boundSubstitution
+                            }
```
Comment (ttuegel) [inline kore/src/Kore/Step/Simplification/Exists.hs:205]: I think this is a little easier to read:

```suggestion
                matched <-
                    normalized
                        { Conditional.substitution = boundSubstitution }
                    & matchesToVariableSubstitution sideCondition variable
                    & lift
```

File: kore/test/Test/Kore/Builtin/Map.hs [hunk 3]
```diff
@@ -1435,10 +1439,12 @@ test_inKeys =
...
-                            (guard . not) (isDefinedPattern termKey)
+                            (guard . not)
+                                (SideCondition.isDefined SideCondition.top termKey)
                             pure (makeCeilPredicate termKey)
                         , do
-                            (guard . not) (isDefinedPattern termMap)
+                            (guard . not)
+                                (SideCondition.isDefined SideCondition.top termMap)
                             pure (makeCeilPredicate termMap)
                         ]
```
Comment (ttuegel) [inline kore/test/Test/Kore/Builtin/Map.hs:1449]: Let's extract a helper function for these list elements.

**Review by ttuegel — CHANGES_REQUESTED** [review]

**Review by andreiburdusa — APPROVED** [review]

**Review by ttuegel — APPROVED** [review]

**Pull request merged**